### PR TITLE
Patch 1

### PR DESCRIPTION
**Pull request by NisargPipaliya**
# Changes made in PR
- Update the code as per reviews received
   -  Applied OOP principle
   - Added Lombok support
   -  Added Google Formatter

### CODE REVIEW

File: src/test/java/updated/game/abstractClasses/Gamer.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.List;
+
+@Getter
+public abstract class Gamer {
+    protected String gamerName;
+    @Setter
+    public char symbol;
+    public int currentScore;
+    public int numberOfMoves;
+
```
Comment (Maitri-Shah24) [inline src/test/java/updated/game/abstractClasses/Gamer.java:15]: Can make currentScore and numberOfMoves private variables and make their getter and setter public methods.

File: src/test/java/updated/game/abstractClasses/Gamer.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    }
+    public void printDetails(){
+        System.out.println("******************************");
+        System.out.printf("%25s", "Details Received are as: \n");
+        System.out.printf("Username: %s\n", this.gamerName);
+        System.out.printf("Symbol Selected by %s: %c\n", this.gamerName, this.symbol);
+        System.out.println("******************************");
+
+    }
+
```
Comment (Maitri-Shah24) [inline src/test/java/updated/game/abstractClasses/Gamer.java:30]: Instead of writing PrintDetails here can can put this method in separate utility class that handles display related logic

File: src/test/java/updated/game/abstractClasses/Gamer.java (hunk 1 above)
Comment (NisargPipaliya) [reply src/test/java/updated/game/abstractClasses/Gamer.java:15]: If i make it private, then i can't inherit them.

File: src/test/java/updated/game/abstractClasses/Board.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+@Getter
+public abstract class Board {
+    public static  int BOARD_SIZE = 3;
```
Comment (Maitri-Shah24) [inline src/test/java/updated/game/abstractClasses/Board.java:13]: BOARD_SIZE can be entered by user so that it can be extensible when you want to implement nXn TicTcToe

File: src/test/java/updated/game/abstractClasses/Board.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    public boolean checkCollision(int index){
+        return !this.emptyCells.contains(index);
+    }
+  
+    public void userMove(Gamer gamer, int row, int col){
+        gamer.incrementNumberOfMoves();
+        this.board.get(row).set(col, gamer.getSymbol());
+    }
+
+    public boolean checkDiagonal() {
```
Comment (Maitri-Shah24) [inline src/test/java/updated/game/abstractClasses/Board.java:58]: The diagonal check logic is a bit hard to follow and specific to a 3x3 grid. To support an n x n grid, you could generalize the diagonal logic for variable sizes.

File: src/test/java/updated/game/abstractClasses/Board.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            if(result){
+                System.out.println("In Check vertical: " + currentSymbol);
+                this.winningSym = currentSymbol;
+                break;
+            }
+        }
+        return this.winningSym != ' ';
+    }
+
+    public boolean checkHorizontal() {
```
Comment (Maitri-Shah24) [inline src/test/java/updated/game/abstractClasses/Board.java]: Horizontal and vertical checks have similar logic. You can reduce redundancy by creating a common method and applying it in both cases.

File: src/test/java/updated/game/abstractClassImpl/EasyComputer.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+import updated.game.abstractClasses .Game;
+import updated.game.abstractClasses.Gamer;
+
+import java.util.List;
+
+public class EasyComputer extends Gamer {
+
+    @Override
+    public void getSymbolFromUser() {
```
Comment (Maitri-Shah24) [inline src/test/java/updated/game/abstractClassImpl/EasyComputer.java:15]: Can avoid this unimplemented method by providing defualt implementation.

File: src/test/java/updated/game/abstractClasses/Board.java (hunk 3 above)
Comment (NisargPipaliya) [reply src/test/java/updated/game/abstractClasses/Board.java:13]: I was thinking, as it's an abstract class so let me keep 3 by default, then i would have a separate implementation of class where user is given choice to enter the size of board he/she is willing to play.

**Review by Maitri-Shah24 — CHANGES_REQUESTED** [review]
Try to make your code extensible to nXn as in some of the methods it is confied to 3X3.

File: src/test/java/updated/game/abstractClasses/Gamer.java (hunk 1 above)
Comment (Maitri-Shah24) [reply src/test/java/updated/game/abstractClasses/Gamer.java:15]: Okay as you are using it to increment values it's fine. But it's better to make variables private and implementing their getters and setters as public method.

File: src/test/java/updated/game/abstractClasses/Board.java (hunk 3 above)
Comment (Maitri-Shah24) [reply src/test/java/updated/game/abstractClasses/Board.java:13]: Keeping the default BOARD_SIZE in the abstract class is fine for a base implementation. However, making it configurable in the derived class where the user can choose their board size would make the game more flexible and align with object-oriented principles.

File: src/test/java/updated/game/abstractClasses/Gamer.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+@Getter
+public abstract class Gamer {
+    protected String gamerName;
+    @Setter
+    public char symbol;
+    public int currentScore;
+    public int numberOfMoves;
```
Comment (shreya6323) [inline src/test/java/updated/game/abstractClasses/Gamer.java]: Is it possible to use list size for tracking the number of moves rather than having a separate data member ?

File: src/test/java/updated/game/abstractClasses/Game.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+import lombok.Getter;
+
+@Getter
+public abstract class Game {
+    protected int numberOfMatches;
+    protected boolean playGame;
+    protected Board board;
+
+    public Game(){
+        numberOfMatches = 1;
```
Comment (shreya6323) [inline src/test/java/updated/game/abstractClasses/Game.java:11]: How about incrementing numberofMatches at the end of the game when you askForRematch ?

File: src/test/java/updated/game/abstractClassImpl/TicTacToe.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        super();
+        human = new Human();
+        computer = new EasyComputer();
+        this.board = new TicTacToeBoard();
+
+    }
+
+    @Override
+    public void printInstructions() {
+        System.out.println("******************************");
```
Comment (shreya6323) [inline src/test/java/updated/game/abstractClassImpl/TicTacToe.java]: Create a method print() for code readability
Can use forloop

File: src/test/java/updated/game/abstractClassImpl/TicTacToe.java [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+        //
+        System.out.println("⚠️Press Q to Quit on going old.game.⚠️ \n\n");
+        System.out.println("The board will be as");
+        StringBuilder cliResult = new StringBuilder();
+        for (int i = 0; i < Board.BOARD_SIZE; i++) {
+            for (int j = 0; j < Board.BOARD_SIZE; j++) {
+                if (j != (Board.BOARD_SIZE - 1))
+                    cliResult.append("X").append("|");
+                else
+                    cliResult.append('X').append("\n");
```
Comment (shreya6323) [inline src/test/java/updated/game/abstractClassImpl/TicTacToe.java]: Can use constant

File: src/test/java/updated/game/abstractClassImpl/TicTacToe.java [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    @Override
+    public void play() {
+        this.printInstructions();
+        do{
+            System.out.println("\n\nLet's Begin \uD83C\uDFB2");
+            System.out.println("\n\nCurrent Status");
+            this.board.printBoard();
+
+            while (!isGameOver(human.getNumberOfMoves(),computer.getNumberOfMoves())) {
+                System.out.printf("========== %s's Turn ==========\n", human.getGamerName());
```
Comment (shreya6323) [inline src/test/java/updated/game/abstractClassImpl/TicTacToe.java]: Acc. to Tic-Tac-Toe, X will get the first chance

File: src/test/java/updated/game/abstractClassImpl/TicTacToe.java [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+                List<Integer> rowAndCol = human.getInputFromUser(this);
+                board.userMove(human, rowAndCol.get(0), rowAndCol.get(1));
+                System.out.println("Board");
+                this.board.printBoard();
+                if (isGameOver(human.getNumberOfMoves(),computer.getNumberOfMoves()) || ((human.getNumberOfMoves() + computer.getNumberOfMoves()) == (TicTacToeBoard.BOARD_SIZE * TicTacToeBoard.BOARD_SIZE)) ) {
+                    System.out.println("IN IF");
+                    break;
+                }
+                System.out.println("========== Computer's Turn ==========");
+                rowAndCol = computer.getInputFromUser(this);
```
Comment (shreya6323) [inline src/test/java/updated/game/abstractClassImpl/TicTacToe.java]: How about incorporating common logic from user and computer move into different method

File: src/test/java/updated/game/abstractClassImpl/TicTacToe.java [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+                this.board.userMove(computer, rowAndCol.getFirst(), rowAndCol.get(1));
+                System.out.println("Board");
+                this.board.printBoard();
+                System.out.println("\n\n");
+            }
+
+            char winningSym = board.getWinningSym();
+            if (winningSym == human.getSymbol() ) {
```
Comment (shreya6323) [inline src/test/java/updated/game/abstractClassImpl/TicTacToe.java]: Can create different method for winner declaration

File: src/test/java/updated/game/abstractClasses/Game.java (hunk 8 above)
Comment (NisargPipaliya) [reply src/test/java/updated/game/abstractClasses/Game.java:11]: Done

File: src/test/java/updated/game/abstractClasses/Gamer.java (hunk 7 above)
Comment (NisargPipaliya) [reply src/test/java/updated/game/abstractClasses/Gamer.java]: Done

Comment (NisargPipaliya) [reply src/test/java/updated/game/abstractClasses/Gamer.java]: Done

File: src/test/java/updated/game/abstractClassImpl/TicTacToe.java (hunk 9 above)
Comment (NisargPipaliya) [reply src/test/java/updated/game/abstractClassImpl/TicTacToe.java]: Done

**Review by Maitri-Shah24 — APPROVED** [review]

**Review by shreya6323 — APPROVED** [review]
Approved !
